File: server/src/routes/push.ts
import { Hono } from 'hono';
import { authMiddleware } from '../auth';
import { r2ConfigForWorld, presignedPutUrl, r2HeadObject, itemKeyToObjectKey } from '../r2';
import { db } from '../db';
import { log } from '../logger';
import type { AuthContext } from '../auth';

type Variables = { auth: AuthContext };

export const pushRouter = new Hono<{ Variables: Variables }>();

// POST /push/init — register push intent, get presigned upload URL
pushRouter.post('/init', authMiddleware(2), async (c) => {
  const auth = c.get('auth');
  const body = await c.req.json<{ key: string; version: string }>();
  const { key, version } = body;

  if (!key || !/^[\w\-:]+$/i.test(key)) {
    return c.json({ error: 'invalid key' }, 400);
  }
  if (!version) {
    return c.json({ error: 'version required' }, 400);
  }

  log.debug(`push/init: world=${auth.world} key=${key} version=${version}`);

  const existing = db.prepare(
    `SELECT 1 FROM items WHERE world = ? AND key = ? AND version = ?`
  ).get(auth.world, key, version);

  if (existing) {
    return c.json({ error: 'version already exists' }, 409);
  }

  const objectKey = itemKeyToObjectKey(key, version);
  const uploadUrl = await presignedPutUrl(r2ConfigForWorld(auth.world), objectKey);

  return c.json({ upload_url: uploadUrl, object_key: objectKey });
});

// POST /push/confirm — verify upload landed in R2, then commit to DB
pushRouter.post('/confirm', authMiddleware(2), async (c) => {
  const auth = c.get('auth');
  const body = await c.req.json<{ key: string; version: string }>();
  const { key, version } = body;

  log.debug(`push/confirm: world=${auth.world} key=${key} version=${version}`);

  const objectKey = itemKeyToObjectKey(key, version);
  const exists = await r2HeadObject(r2ConfigForWorld(auth.world), objectKey);
  if (!exists) {
    return c.json({ error: 'object not found in R2 — upload may not have completed' }, 400);
  }

  db.prepare(
    `INSERT OR IGNORE INTO items (world, key, version, created_at) VALUES (?, ?, ?, unixepoch())`
  ).run(auth.world, key, version);

  return c.json({ ok: true });
});

File: c2server/routes/push.go
package routes

import (
	"cacher/c2server/internal"
	"regexp"
	"time"

	"github.com/gofiber/fiber/v2"
)

var validItemKey = regexp.MustCompile(`^[\w\-:]+$`)

func SetupPush(app *fiber.App) {
	g := app.Group("/push")
	g.Post("/init", internal.AuthMiddleware(2), pushInit)
	g.Post("/confirm", internal.AuthMiddleware(2), pushConfirm)
}

func pushInit(c *fiber.Ctx) error {
	auth := internal.GetAuth(c)

	var body struct {
		Key     string `json:"key"`
		Version string `json:"version"`
	}
	if err := c.BodyParser(&body); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "invalid request body"})
	}
	if !validItemKey.MatchString(body.Key) {
		return c.Status(400).JSON(fiber.Map{"error": "invalid key"})
	}
	if body.Version == "" {
		return c.Status(400).JSON(fiber.Map{"error": "version required"})
	}

	internal.LogDebug("push/init: world=%s key=%s version=%s", auth.World, body.Key, body.Version)

	var exists int
	err := internal.DB.QueryRow(
		"SELECT 1 FROM items WHERE world = ? AND key = ? AND version = ?",
		auth.World, body.Key, body.Version,
	).Scan(&exists)
	if err == nil {
		return c.Status(409).JSON(fiber.Map{"error": "version already exists"})
	}

	objectKey := internal.ItemKeyToObjectKey(body.Key, body.Version)
	uploadURL, err := internal.PresignedPutURL(auth.World, objectKey, 15*time.Minute)
	if err != nil {
		internal.LogError("push/init: presign failed: %v", err)
		return c.Status(500).JSON(fiber.Map{"error": "could not generate upload URL"})
	}

	return c.JSON(fiber.Map{"upload_url": uploadURL, "object_key": objectKey})
}

func pushConfirm(c *fiber.Ctx) error {
	auth := internal.GetAuth(c)

	var body struct {
		Key     string `json:"key"`
		Version string `json:"version"`
	}
	if err := c.BodyParser(&body); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "invalid request body"})
	}

	internal.LogDebug("push/confirm: world=%s key=%s version=%s", auth.World, body.Key, body.Version)

	objectKey := internal.ItemKeyToObjectKey(body.Key, body.Version)
	exists, err := internal.R2HeadObject(auth.World, objectKey)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "could not verify upload"})
	}
	if !exists {
		return c.Status(400).JSON(fiber.Map{"error": "object not found in R2 — upload may not have completed"})
	}

	internal.DB.Exec(
		"INSERT OR IGNORE INTO items (world, key, version, created_at) VALUES (?, ?, ?, unixepoch())",
		auth.World, body.Key, body.Version,
	)

	return c.JSON(fiber.Map{"ok": true})
}
